Allow using zarr v3

Currently, zarr dependency is pinned to >=2.18, <3 until either test `tests.plugins.xcube.processors.test_mldataset.MultiLevelDatasetProcessorTest` is adjusted or fsspec's memory filesystem is updated to be async.

See also: https://github.com/dask/dask/issues/11691

With zarr 3, the test fails as follows:

```
________________ MultiLevelDatasetProcessorTest.test_preprocess ________________

self = <tests.plugins.xcube.processors.test_mldataset.MultiLevelDatasetProcessorTest testMethod=test_preprocess>

    def setUp(self):
        self._delete_files()
        self.fs.mkdir(self.levels_dir)
    
        level_datasets = make_cube_levels(
            self.num_levels, self.x_size, self.y_size, self.time_size
        )
        for level, dataset in enumerate(level_datasets):
>           dataset.to_zarr(f"{self.levels_dir}/{level}.zarr", write_empty_chunks=False)

tests/plugins/xcube/processors/test_mldataset.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/xarray/core/dataset.py:2622: in to_zarr
    return to_zarr(  # type: ignore[call-overload,misc]
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/xarray/backends/api.py:2194: in to_zarr
    zstore = backends.ZarrStore.open_group(
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/xarray/backends/zarr.py:703: in open_group
    ) = _get_open_params(
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/xarray/backends/zarr.py:1792: in _get_open_params
    zarr_group = zarr.open_group(store, **open_kwargs)
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/zarr/_compat.py:43: in inner_f
    return f(*args, **kwargs)
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/zarr/api/synchronous.py:524: in open_group
    sync(
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/zarr/core/sync.py:142: in sync
    raise return_result
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/zarr/core/sync.py:98: in _runner
    return await coro
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/zarr/api/asynchronous.py:800: in open_group
    store_path = await make_store_path(store, mode=mode, storage_options=storage_options, path=path)
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/zarr/storage/_common.py:305: in make_store_path
    store = FsspecStore.from_url(
/opt/hostedtoolcache/Python/3.13.1/x64/lib/python3.13/site-packages/zarr/storage/_fsspec.py:182: in from_url
    return cls(fs=fs, path=path, read_only=read_only, allowed_exceptions=allowed_exceptions)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <FsspecStore(MemoryFileSystem, /xrlint-test.levels/0.zarr)>
fs = <fsspec.implementations.memory.MemoryFileSystem object at 0x7f10f406c050>
read_only = False, path = '/xrlint-test.levels/0.zarr'
allowed_exceptions = (<class 'FileNotFoundError'>, <class 'IsADirectoryError'>, <class 'NotADirectoryError'>)

    def __init__(
        self,
        fs: AsyncFileSystem,
        read_only: bool = False,
        path: str = "/",
        allowed_exceptions: tuple[type[Exception], ...] = ALLOWED_EXCEPTIONS,
    ) -> None:
        super().__init__(read_only=read_only)
        self.fs = fs
        self.path = path
        self.allowed_exceptions = allowed_exceptions
    
        if not self.fs.async_impl:
>           raise TypeError("Filesystem needs to support async operations.")
E           TypeError: Filesystem needs to support async operations.

```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -46,6 +46,7 @@
 
 ### Other changes
 
+- XRLint now works with zarr >=2,<3 and zarr >=3.0.2
 - Added more tests so we finally reached 100% coverage.
 - New `PluginMeta.docs_url` property.
 

--- a/environment.yml
+++ b/environment.yml
@@ -1,4 +1,4 @@
-name: xrlint-zarr3
+name: xrlint
 channels:
   - conda-forge
 dependencies:
@@ -24,4 +24,5 @@ dependencies:
   - pandas
   - netcdf4
   - numpy
-  - zarr >=2.18
+  # actually zarr >=2.18,<3.0 | >=3.0.2
+  - zarr

--- a/environment.yml
+++ b/environment.yml
@@ -24,5 +24,4 @@ dependencies:
   - pandas
   - netcdf4
   - numpy
-  # actually zarr >=2.18,<3.0 | >=3.0.2
-  - zarr
+  - zarr >=2.18,!=3.0.0,!=3.0.1

--- a/environment.yml
+++ b/environment.yml
@@ -1,4 +1,4 @@
-name: xrlint
+name: xrlint-zarr3
 channels:
   - conda-forge
 dependencies:
@@ -24,4 +24,4 @@ dependencies:
   - pandas
   - netcdf4
   - numpy
-  - zarr >=2.18,<3  # tests fail with zarr 3+
+  - zarr >=2.18

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -77,7 +77,8 @@ dev = [
   "netcdf4",
   "numpy",
   "pandas",
-  "zarr >=2.18,<3",
+  # actually zarr >=2.18,<3.0 | >=3.0.2
+  "zarr",
 ]
 doc = [
   "mkdocs",

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -77,8 +77,7 @@ dev = [
   "netcdf4",
   "numpy",
   "pandas",
-  # actually zarr >=2.18,<3.0 | >=3.0.2
-  "zarr",
+  "zarr>= 2.18, != 3.0.0, != 3.0.1",
 ]
 doc = [
   "mkdocs",

--- a/tests/plugins/xcube/processors/test_mldataset.py
+++ b/tests/plugins/xcube/processors/test_mldataset.py
@@ -50,7 +50,7 @@ class MultiLevelDatasetProcessorTest(TestCase):
             self.num_levels, self.x_size, self.y_size, self.time_size
         )
         for level, dataset in enumerate(level_datasets):
-            dataset.to_zarr(f"{self.levels_dir}/{level}.zarr", write_empty_chunks=False)
+            dataset.to_zarr(f"{self.levels_dir}/{level}.zarr")
 
         with self.fs.open(self.meta_path, mode="wt") as stream:
             json.dump(self.meta_content, stream, indent=2)

--- a/tests/test_linter.py
+++ b/tests/test_linter.py
@@ -77,7 +77,7 @@ class LinterValidateWithConfigTest(TestCase):
 
     def test_config_with_file_not_found(self):
         linter = new_linter({"rules": {"no-empty-attrs": 2}})
-        result = linter.validate("cube.zarr")
+        result = linter.validate("cube.nc")
         self.assert_result_ok(result, "No such file or directory:")
 
     def test_no_config(self):

